Base Delaunay fitter on scipy.interpolate.griddata

ludwigschwardt opened the pull request:
The old delaunay module is now completely gone, having been deprecated
for a while. The ideal replacement is `scipy.interpolate.griddata` and its
underlying classes, which is fast and robust.

Also remove Delaunay2DGridFit which serves no real purpose anymore.

ludwigschwardt committed:
Replace prints with logging

Also define the logger in delaunay where it was used without definition.

ludwigschwardt committed:
Remove Delaunay2DGridFit

This class was never really used as Delaunay is a natural scatter
fitter. It was only there because the nn_interpolator option in the
old delaunay module had two modes: interp(x0, x1) for unstructured
data and interp[x0, x1] for gridded data. Both could be handled by
the former, but this had issues in the old module, resulting in the
special case.

ludwigschwardt committed:
Update Delaunay to use matplotlib.tri

The old delaunay module is now completely gone, having been deprecated
for a while. There are currently two possible replacements found in
scipy.spatial and matplotlib.tri. I chose the latter since it also comes
with interpolation routines.

Fixes issue #2.

ludwigschwardt committed:
Now also depend on matplotlib for Delaunay

If scikit-learn and skimage can do it, why not us?

bmerry commented on scikits/fitting/__init__.py:
```diff
@@ -77,14 +77,17 @@
     except ImportError:
         def test():
-            print('Could not load nose. Unit tests not available.')
+            logger.debug('Could not load nose. Unit tests not available.')
```
Will this message ever be visible i.e. is logging actually configured at this point?

bmerry commented on setup.py:
```diff
@@ -52,7 +52,7 @@ def configuration(parent_package='', top_path=None):
                        'License :: OSI Approved :: BSD License',
                        'Topic :: Scientific/Engineering'],
           configuration=configuration,
-          install_requires=['numpy', 'scipy'],
```
Should put a minimum version on matplotlib that provides the required functionality.

ludwigschwardt replied:
I have a minimum version on scipy now which simplifies other bits too.

ludwigschwardt committed:
Redo Delaunay again based on scipy.interpolate

I somehow missed scipy.interpolate.griddata all these years...

This helper function does Delaunay triangulation via Qhull and cubic or
linear interpolation all rolled into one. Rewrite Delaunay2DScatterFit
to be based on this instead, as it is orders of magnitude faster than
matplotlib.tri (probably because it is compiled C code). Actually use
the underlying interpolator classes as this allows the fitting and
evaluation stages to be separated (useful for fitting once and
evaluating many times).

ludwigschwardt committed:
Depend on scipy >= 0.9.0

This ensures that the Delaunay interpolators are available in
scipy.interpolate. In addition, remove all checks for scipy 0.7.0
involving PiecewisePolynomial and Rbf interpolators as this will
automatically be satisfied. Coincidentally this picked up a bug
where PiecewisePolynomial1DFit wasn't tested anymore as the old
underlying class disappeared from the latest scipy.

The dependency is not particularly onerous as scipy 0.9.0 dates
back to 2011... Also, there is no need for matplotlib anymore.

ludwigschwardt committed:
Upgrade missing nose check to warning

bmerry commented on README.rst:
```diff
@@ -49,8 +49,6 @@ Grid fitters
 
 - ``GridFit``: Abstract base class for grid fitters
-- ``Delaunay2DGridFit``: Interpolate scalar function defined on 2-D grid,
-  based on Delaunay triangulation
```
Should the README maybe have a note about backwards-incompatible changes, just in case someone is actually using it.

ludwigschwardt replied:
I'm not too worried about this class - no-one within SKA ever used it. If someone asks about it, it will be a useful market research exercise :-)

ludwigschwardt replied:
Thinking about it again, I'll add a changelog section in the long description...

ludwigschwardt replied:
This will only be in the next PR readying for the release though.

bmerry commented on scikits/fitting/__init__.py:
```diff
@@ -77,14 +77,17 @@
     except ImportError:
         def test():
-            print('Could not load nose. Unit tests not available.')
+            logger.warning('Could not load nose. Unit tests not available.')
```
I'm not sure exactly how this is intended to be used, but if the logging module hasn't been set up all the user will see is a message that there are no handlers (or possibly nothing at all, I forget). Maybe the `warnings` module is a better fit here?

ludwigschwardt replied:
Point taken! Both instances of logging replaced with warnings.warn().

bmerry commented on scikits/fitting/delaunay.py:
```diff
@@ -1,70 +1,58 @@
-    interp_type : {'nn'}, optional
-        String indicating type of interpolation (only 'nn' currently supported)
+    interp_type : {'cubic', 'linear', 'nearest'}, optional
+        String indicating type of interpolation
```
Is `nearest` different to `nn`?

ludwigschwardt replied:
Yes. `nn` is "natural neighbours", while `nearest` is "nearest neighbours". The former is not easily available anymore. The latter does not actually involve Delaunay triangulation but is provided in scipy alongside the other two. I won't tell if you won't :-)

bmerry commented on scikits/fitting/delaunay.py:
```diff
@@ -1,70 +1,58 @@
     jitter : bool, optional
         True to add small amount of jitter to *x* to make degenerate
-        triangulation unlikely
+        triangulation unlikely (generally not needed with Qhull back-end)
```
Is the comment about Qhull still relevant now that you have only the scipy backend?

ludwigschwardt replied:
What I meant here is that people should be discouraged from setting `jitter=True` as it serves very little purpose now that the triangulation is more robust via Qhull. On the other hand it also does very little harm, which is why it is a bit meh.

bmerry replied:
Maybe just say " (generally not needed)" then, since the not-needed-ness is unconditional.

ludwigschwardt replied:
Good. I now see how you read it.

bmerry commented on scikits/fitting/tests/test_delaunay.py:
```diff
@@ -27,66 +27,43 @@ def setUp(self):
+        assert_almost_equal(testy, self.testy, decimal=10)
+        # Nearest-neighbour interpolation has no outside value
+        # assert_almost_equal(outsidey, self.outsidey, decimal=10)
+
```
Not something you really need to fix in this pull request, but an interpolation test where all the values being interpolated between are equal doesn't seem particularly strong...

ludwigschwardt replied:
For me the test is more about API verification and getting to the magical 100% coverage ;-)

ludwigschwardt committed:
Replace logging with warnings

The warnings module is more suitable for these two warnings, plus its
output can still be redirected to the usual logging if needed.

ludwigschwardt committed:
Fix wording in docstring

The meaning is this: people should be discouraged from setting
jitter=True as it serves very little purpose now that the triangulation
is based on Qhull and therefore much more robust. On the other hand
it also does very little harm, which is why it is a bit meh.
